src/shillelagh/adapters/api/google_analytics.py

```diff
@@ -0,0 +1,218 @@
+from typing import Tuple, Any, Dict, List, Iterator, Optional
+from shillelagh.filters import Filter
+from shillelagh.fields import Field, String
+from shillelagh.typing import RequestedOrder, Row
+from apiclient.discovery import build
+from sqlparse import parse
+from sqlparse.sql import Identifier, IdentifierList, Where, Comment
+
+from oauth2client.service_account import ServiceAccountCredentials
+
+# all apis: https://developers.google.com/analytics
+# this implements adapter for core reporting api v4
+# later port it to upcoming GA4 property as this will be deprecated by google
+
+"""
+TODO: checkout apsw doc for pagination with virtual cursor so we query GA with correct offset
+
+TODO: fill all items in ALL_DIMESIONS and ALL_METRICS in below lists
+
+TODO: in google_analytics dialect find a way to correctly decode http encoded content saved by superset
+
+TODO: currently date ranges for query need to be setup when db is added, modify shillelagh to give adapter a hook to modify query which will remove those ranges from the query but use it when querying GA only.
+        this is required because date ranges on GA are like "7daysAgo", "today" which are not valid in sqlite
+
+TODO: currently all dimesions and metrics are treated as string, use appropriate data type
+"""
+
+# source: https://ga-dev-tools.web.app/dimensions-metrics-explorer/
+ALL_DIMENSIONS = [
+    #User
+    "ga:userType",
+    "ga:sessionCount",
+    "ga:daysSinceLastSession",
+    "ga:userDefinedValue",
+    "ga:userBucket",
+
+    #Session
+    "ga:sessionDurationBucket",
+
+    #Traffic sources
+    "ga:referralPath",
+    "ga:fullReferrer",
+    "ga:campaign",
+    "ga:source",
+    "ga:medium",
+    "ga:sourceMedium",
+    "ga:keyword",
+    "ga:adContent",
+    "ga:adContent",
+    "ga:hasSocialSourceReferral",
+    "ga:campaignCode",
+]
+ALL_METRICS = [
+    #User
+    "ga:users",
+    "ga:newUsers",
+    "ga:percentNewSessions",
+    "ga:1dayUsers",
+    "ga:7dayUsers",
+    "ga:14dayUsers",
+    "ga:28dayUsers",
+    "ga:30dayUsers",
+    "ga:sessionsPerUser",
+
+    #Session
+    "ga:sessions",
+    "ga:bounces",
+    "ga:bounceRate",
+    "ga:sessionDuration",
+    "ga:avgSessionDuration",
+    "ga:uniqueDimensionCombinations",
+    "ga:hits",
+
+    #Traffic Sources
+    "ga:organicSearches"
+]
+
+SCOPES = ["https://www.googleapis.com/auth/analytics.readonly"]
+
+class GoogleAnalytics(Adapter):
+    columns: Dict[str, Field] = {}
+
+    def __init__(
+        self,
+        uri,
+        service_account_info: str = None,
+        view_id=None,
+        operation: str = None,
+        parameters=Optional[Tuple[Any, ...]],
+        date_range = None,
+    ):
+        super().__init__()
+        self.view_id = view_id
+        self.operation = operation
+        self.parameters = parameters
+        self.date_range = date_range
+
+        credentials = ServiceAccountCredentials.from_json_keyfile_dict(
+            service_account_info, SCOPES
+        )
```

Comment:
We should probably reuse https://github.com/betodealmeida/shillelagh/blob/ee152b531a3e462c9f078b12eee4fb17fd690fff/src/shillelagh/adapters/api/gsheets/lib.py#L211 here, so other authentication methods are also supported.